### ISSUE TITLE
Easy execution of tasks from within Python itself

### ISSUE THREAD

**Issue by bitprophet** [pyinvoke/invoke#170]
Until now the focus has been mostly on CLI execution, but calling programmatically has always been on the table - couldn't find a good ticket so this must be it :) (#112 is sort of a feature request for this but it's more of a support thread, I'll roll it into here instead.)

Currently there are no great options for calling a task from another task:
- The "main" way to execute presently is `Executor(<Collection obj>).execute('name')`, which requires one to have access to the right `Collection` instance.
  - Especially if that `Collection` is the one the task is itself a part of, this is basically impossible with the current API.
- Executing a task that lives in an imported collection is easier, one can simply do `that_collection['taskname'](args)`, however:
  - it's still a little fugly (if not a ton)
  - more importantly, it has annoying edge cases when using contextualized tasks - the context objects are different from each other so you can't just hand yours into the other task (though this feels like it should be allowed!)

Rambling thoughts about how to smooth over the above problems:
- Having some focus `Collection` when executing is still a must at some point - Fabric's `execute('name')` can't work here because it relies on implicit global state.
- The connection between a task's inner block and the Collection it lives in, must be via the `Context` handed in (or some other argument, but a single forced argument is plenty...).
  - Feels like more unification of (or smoothing over of access to) `Context` and `Collection` might help here.
  - At the very least we need a way to easily call a task with the "right" collection-as-configuration tree - similar to what the CLI mechanisms do now. Which may mean shoving more of that code into `Collection` or `Context`.
- The previous bullet is more fuel for removing non-contextualized tasks (#114)
  - However, the non-contextualized route just means you can't easily call contextualized tasks from non-contextualized ones (w/o making your own `Context` - which is doable). You can always just treat everything like a regular module-of-functions and call other nearby task functions straight up.

Brainstorm of use cases/scenarios:
- Call one top level task from another
- Call a subtask from a top level task
- ~~Call a top level task from a subtask~~ (this doesn't actually make any sense, by definition a subtask has no idea that it IS a subtask, at best it can be expected to know about its siblings)
- Call a subtask from another subtask in the same module (though ideally this is exactly the same as the first bullet)
- Call a top level task as a library call (eg in the shell)
- Call a subtask as a library call
- (Potentially a new ticket) Call one or more top level tasks, handing them the results of the parser - basically, control over the existing hardcoded-ish CLI module.

**Comment by jhermann** [comment]
NOT having read the long thread, I'm still throwing #223 in the ring. :smile:

**Comment by frol** [comment]
Just in case somebody needs a workaround to execute tasks from another tasks right now, I implemented a really simple way to do so in my tasks by passing a root namespace and a helper function into `context`:

```python
from invoke import ctask, Collection
from invoke.executor import Executor

# Define tasks
# ============

@ctask
def test_task1(context, arg1):
    print("TASK1")

@ctask
def test_task2(context):
    print("TASK2 BEGIN")
    context.invoke_execute(
         context,
         'test_task1',
         arg1='value1'
    )
    print("TASK2 END")


# Define and configure root namespace
# ===================================

# NOTE: `namespace` or `ns` name is required!
namespace = Collection(
    test_task,
    # put tasks or nested collections here
)

def invoke_execute(context, command_name, **kwargs):
    """
    Helper function to make invoke-tasks execution easier.
    """
    results = Executor(namespace, config=context.config).execute((command_name, kwargs))
    target_task = context.root_namespace[command_name]
    return results[target_task]

namespace.configure({
    'root_namespace': namespace,
    'invoke_execute': invoke_execute,
})
```

**Comment by pdonorio** [comment]
Sorry for bothering. This issue is 3 years old, I was wondering if we've reached better ways now to call tasks from other tasks. Thanks!

**Comment by bitprophet** [comment]
No, but it remains at or near the top of the priority list; it's definitely something that needs solving before 1.0.

**Comment by TimotheeJeannin** [comment]
Any updates on this ?

**Comment by pdonorio** [comment]
> it's definitely something that needs solving before 1.0.

@bitprophet I see that 1.0 was released :)

**Comment by muhammedabad** [comment]
With the release of 1.0, is the ability to call tasks from within tasks (similar to v1's `execute`) available ?

**Comment by bitprophet** [comment]
Re: @muhammedabad's 2nd question, see #261 - it needs work still!

Re: this ticket and 1.0: hey, running these projects isn't a cakewalk! ;) I judged it was better to get this & related projects above-board and on semver, than to continually punt until it was perfect.

My gut says we can definitely implement this in a backwards compatible manner, since it's likely to end up implemented as a new method on `Context` (plus related changes to things like `Executor`). So it should appear in some 1.x feature release!

**Comment by SamuelMarks** [comment]
Any updates?

I've [written a set of configuration management tools around Fabric (and Apache Libcloud)](https://github.com/offscale).

But can't upgrade it to be Python 2.7-3+ compatible until I upgrade Fabric. But that requires `execute` at a minimum.

Roadmap?

**Comment by jmsuzuki** [comment]
> Any updates?
> 
> I've [written a set of configuration management tools around Fabric (and Apache Libcloud)](https://github.com/offscale).
> 
> But can't upgrade it to be Python 2.7-3+ compatible until I upgrade Fabric. But that requires `execute` at a minimum.
> 
> Roadmap?

I'm in the same boat.  I need execute to migrate from python 2.7 to 3+

**Comment by rectalogic** [comment]
> Just in case somebody needs a workaround to execute tasks from another tasks right now, I implemented a really simple way to do so in my tasks by passing a root namespace and a helper function into `context`:

To get this to work with fabric, and honor the executed tasks host list, I had to `from fabric.main import program` and pass `program.core` to the `Executor`:

```python
from invoke import Collection
from fabric import task
from fabric.executor import Executor
from fabric.main import program

# Define tasks
# ============


@task(hosts=["localhost"])
def test_task1(context, arg1):
    print("TASK1", context)


@task
def test_task2(context):
    print("TASK2 BEGIN", context)
    context.invoke_execute(
         context,
         'test_task1',
         arg1='value1'
    )
    print("TASK2 END")


# Define and configure root namespace
# ===================================

# NOTE: `namespace` or `ns` name is required!
namespace = Collection(
    test_task1,
    test_task2,
    # put tasks or nested collections here
)


def invoke_execute(context, command_name, **kwargs):
    """
    Helper function to make invoke-tasks execution easier.
    """
    results = Executor(namespace, config=context.config, core=program.core).execute((command_name, kwargs))
    target_task = context.root_namespace[command_name]
    return results[target_task]


namespace.configure({
    'root_namespace': namespace,
    'invoke_execute': invoke_execute,
})
```

```sh-session
$ fab test-task2
('TASK2 BEGIN', <Context: <Config: {'root_namespace': <Collection None: test-task1, test-task2>, 'tasks': {'search_root': None, 'collection_name': 'fabfile', 'dedupe': True, 'auto_dash_names': True}, 'run': {'shell': '/bin/bash', 'hide': None, 'pty': False, 'encoding': None, 'in_stream': None, 'replace_env': True, 'echo': False, 'warn': False, 'echo_stdin': None, 'watchers': [], 'env': {}, 'out_stream': None, 'err_stream': None, 'fallback': True}, 'timeouts': {'connect': None}, 'sudo': {'password': None, 'prompt': '[sudo] password: ', 'user': None}, 'inline_ssh_env': False, 'port': 22, 'load_ssh_configs': True, 'user': 'cureatr', 'ssh_config_path': None, 'invoke_execute': <function invoke_execute at 0x7f1993b41f50>, 'connect_kwargs': {}, 'forward_agent': False, 'gateway': None, 'runners': {'remote': <class 'fabric.runners.Remote'>, 'local': <class 'invoke.runners.Local'>}}>>)
('TASK1', <Connection host=localhost>)
TASK2 END
```

**Comment by breisig** [comment]
Any update on this?

**Comment by SamuelMarks** [comment]
I've switched to the Python 3 compatible fab-classic in the meantime.

**Comment by SamuelMarks** [comment]
Another year has gone by… any update?

**Comment by christian-intra2net** [comment]
Also greatly missing this feature. I have some subtasks that I would like to either call directly from the command line or from other tasks.

**Comment by jcw** [comment]
Would this feature also help with the following scenario I'm after?

* an `inv live` task launches the `fswatch` utility to track file changes
* it then continuously watches for changes by responding to output lines
* on a change, launch another task, which recompiles and runs the target app
* rinse and repeat, IOW my goal here is to use invoke as a live-coding mechanism

So the question is whether invoke can be made to _repeatedly_ re-launch a task.
Or is there some other way? (with apologies if I'm mis-reading the gist of this issue)

**Comment by SamuelMarks** [comment]
@jcw - I think you want https://github.com/gorakhargosh/watchdog

**Comment by flavius** [comment]
Three years later?
